### PR TITLE
Fix drafts being allowed outside of Drafts/Trash and non-drafts being allowed inside Drafts

### LINKED ISSUE #3241: Draft moving is inconsistent
We generally do not allow drafts to be moved between folders. However:
 - It is currently still possible to drag draft to another folder
 - There's no button to restore the draft from trash

# Test notes
- [ ] Select a single, normal email
    - [ ] (Desktop/Web) Verify that nothing happens when this email is dragged into the Drafts folder.
    - [ ] (Desktop/Web) Press V. Verify Drafts is not listed.
    - [ ] Click the move button. Verify Drafts is not listed.
- [ ] Select multiple normal emails
    - [ ] (Desktop/Web) Verify that nothing happens when these emails are dragged into the Drafts folder.
    - [ ] (Desktop/Web) Press V. Verify Drafts is not listed.
    - [ ] Click the move button. Verify Drafts is not listed.
- [ ] Select a draft in the drafts folder
    - [ ] (Desktop/Web) Verify that nothing happens when this email is dragged into a folder that is not Trash.
    - [ ] (Desktop/Web) Press V. Verify that only Trash is listed.
    - [ ] Click the move button. Verify that only Trash is listed.
- [ ] Select a draft in the trash.
    - [ ] (Desktop/Web) Verify that nothing happens when this email is dragged into a folder that is not Drafts.
    - [ ] (Desktop/Web) Press V. Verify that only Drafts is listed.
    - [ ] Click the move button. Verify that only Drafts is listed.
 - [ ] Select multiple drafts in the trash.
    - [ ] (Desktop/Web) Verify that nothing happens when these emails are dragged into a folder that is not Drafts.
    - [ ] (Desktop/Web) Press V. Verify that only Drafts is listed.
    - [ ] Click the move button. Verify that only Drafts is listed.
  - [ ] From the trash, select an email that is a draft and one that is not.
    - [ ] (Desktop/Web) Verify dragging these emails does not work.
    - [ ] (Desktop/Web) Press V. Verify that nothing happens.
    - [ ] Verify there is no move button.
  - [ ] On mobile (or with developer tools emulating mobile):
    - [ ] Verify that swiping left on a draft in Drafts does nothing (besides deselecting).
    - [ ] Verify that swiping left on a draft in the Trash recovers it into the Drafts folder.
    - [ ] Verify that swiping left on a normal email in the Trash recovers it into the Inbox folder.
- [ ] On a Premium account, in User settings -> Emails, try to create an inbox rule and verify that Drafts is not listed in the dialog as a potential target folder.

### PR DESCRIPTION
**Pull request by paw-hub**
Ensure people do not move drafts outside of either the drafts folder or trash folder. Also ensure people do not move non-drafts into the drafts folder.

To do this, we do the following:
* If the selection is not a draft, prevent 'Drafts' from being listed in all move button dropdowns.
* If the selection is a draft, prevent anything that is not 'Drafts' or 'Trash' from being listed in all move button dropdowns.
* Do not allow mails to be dragged into disallowed folders.
* Prevent drafts and non-drafts from being moved simultaneously.
* Prevent inbox rules from targeting the drafts folder.

Fixes #3241

### CODE REVIEW

File: src/mail/model/MailUtils.ts [hunk 1]
```diff
@@ -303,6 +293,60 @@ export function markMails(entityClient: EntityClient, mails: Mail[], unread: boo
 	).then(noOp)
 }
 
+/**
+ * Check if all mails in the selection are drafts. If there are mixed drafts and non-drafts or the array is empty, return null. Otherwise, return true if drafts and false if not.
```
Comment (johnbotris) [inline src/mail/model/MailUtils.ts:297]: The documentation on this is unclear, and the name and return value kind of obscure what this function is actually checking. This should return an enum, something like `Mixed`, `Drafts`, `NonDrafts`. I think the empty list case could probably be considered `Mixed`, or just make it an error and require you check before hand? 

Disregarding all of that, it does too much, and we only ever check the return value for `null`, so if all we care about is whether or not it's mixed/empty, then that's all we need to check
```
function containsDraftsAndNonDrafts(mails: ReadonlyArray<Mail>): boolean {
    mails.length === 0
    || (
        mails.some(mail => mail.state === MailState.DRAFT) 
        && mails.some(mail => mail.state !== MailState.DRAFT)
    )
}
````

if we find ourselved caring about the specific cases of all drafts or all mails, then we can just check those individually as needed.

File: src/mail/model/MailUtils.ts [hunk 2]
```diff
@@ -303,6 +293,60 @@ export function markMails(entityClient: EntityClient, mails: Mail[], unread: boo
 	).then(noOp)
 }
 
+/**
```
Comment (johnbotris) [inline src/mail/model/MailUtils.ts:296]: some little tests for these new functions would be good too

File: src/mail/view/MailView.ts [hunk 3]
```diff
@@ -450,8 +450,10 @@ export class MailView implements CurrentView {
 				}
 
 				const filteredFolders = folders.filter(f => f.mails !== mailList)
-				const targetFolders = getSortedSystemFolders(filteredFolders).concat(getSortedCustomFolders(filteredFolders))
-				return targetFolders.map(f => ({
+				return getSortedSystemFolders(filteredFolders)
+					.concat(getSortedCustomFolders(filteredFolders))
+					.filter(folder => allMailsCanBeInsideFolder(selectedMails, folder))
+					.map(f => ({
```
Comment (johnbotris) [inline src/mail/view/MailView.ts:456]: The arrow function body needs to be indented, the formatting made my brain do a 720

File: src/mail/view/MailViewer.ts [hunk 4]
```diff
@@ -678,9 +679,38 @@ export class MailViewer implements Component<MailViewerAttrs> {
 			})
 	}
 
+	private async getAvailableFolders(): Promise<MailFolder[]> {
```
Comment (johnbotris) [inline src/mail/view/MailViewer.ts:682]: Change the name or document. What are the folders available for?

File: src/mail/view/MailViewer.ts [hunk 5]
```diff
@@ -678,9 +679,38 @@ export class MailViewer implements Component<MailViewerAttrs> {
...
+		const folders = await this.viewModel.mailModel.getMailboxFolders(this.viewModel.mail)
+		const filteredFolders = folders.filter(f => f.mails !== this.viewModel.getMailId()[0])
```
Comment (johnbotris) [inline src/mail/view/MailViewer.ts:684]: should be `listIdPart(this.viewModel.getMailId())`

**Review by johnbotris — CHANGES_REQUESTED** [review]

File: src/mail/view/MailViewer.ts (hunk 5 above)
Comment (paw-hub) [reply src/mail/view/MailViewer.ts]: Must have been from the code I extracted. I went and fixed this.

File: src/mail/view/MailViewer.ts (hunk 4 above)
Comment (paw-hub) [reply src/mail/view/MailViewer.ts]: I decided to just not extract this code since it's only used once.

File: src/mail/view/MailView.ts (hunk 3 above)
Comment (paw-hub) [reply src/mail/view/MailView.ts:458]: Oops, fixed. I have no idea how I missed that. Thank goodness for ctrl-s in WebStorm.

File: src/mail/model/MailUtils.ts [hunk 6]
```diff
@@ -303,6 +293,60 @@ export function markMails(entityClient: EntityClient, mails: Mail[], unread: boo
...
+ * Check if all mails in the selection are drafts. If there are mixed drafts and non-drafts or the array is empty, return null. Otherwise, return true if drafts and false if not.
+ * @param mails
+ */
+export function allMailsAreDrafts(mails: readonly Mail[]): boolean | null {
```
[the remark this answers is not in the document]
Comment (paw-hub) [reply src/mail/model/MailUtils.ts]: You're right! I went and fixed that.

File: src/mail/model/MailUtils.ts (hunk 2 above)
Comment (paw-hub) [reply src/mail/model/MailUtils.ts:296]: Tested! The file name's a little long but it should do the job!

**Review by johnbotris — APPROVED** [review]

**Pull request merged**